components/calendar/src/cal/japanese.rs

```diff
@@ -166,27 +169,53 @@ impl JapaneseExtended {
     pub(crate) const DEBUG_NAME: &'static str = "Japanese (historical era data)";
 }
 
+impl DateFieldsResolver for Japanese {
+    type YearInfo = i32;
+
+    #[inline]
+    fn year_info_from_era(&self, era: &str, era_year: i32) -> Result<Self::YearInfo, DateError> {
+        let era = match era {
+            "ce" | "ad" => {
+                return Ok(era_year);
+            }
+            "bce" | "bc" => {
+                return Ok(1 - era_year);
+            }
+            e => e.parse().map_err(|_| DateError::UnknownEra)?,
+        };
+        let (era_start, _next_era_start) = self.japanese_era_range_for(era)?;
+        Ok(era_start.year + era_year - 1)
+    }
+
+    #[inline]
+    fn year_info_from_extended(&self, extended_year: i32) -> Self::YearInfo {
+        extended_year
+    }
+
+    #[inline]
+    fn reference_year_from_month_day(
+        &self,
+        _month_code: types::MonthCode,
+        _day: u8,
+    ) -> Result<Self::YearInfo, DateError> {
+        Ok(Iso::REFERENCE_YEAR)
+    }
+}
+
 impl crate::cal::scaffold::UnstableSealed for Japanese {}
 impl Calendar for Japanese {
     type DateInner = JapaneseDateInner;
     type Year = types::EraYear;
 
-    fn from_codes(
+    fn from_fields(
         &self,
-        era: Option<&str>,
-        year: i32,
-        month_code: types::MonthCode,
-        day: u8,
+        fields: DateFields,
+        options: DateFromFieldsOptions,
     ) -> Result<Self::DateInner, DateError> {
-        let Some((month, false)) = month_code.parsed() else {
-            return Err(DateError::UnknownMonthCode(month_code));
-        };
-
-        if month > 12 {
-            return Err(DateError::UnknownMonthCode(month_code));
-        }
-
-        self.new_japanese_date_inner(era, year, month, day)
+        let builder = ArithmeticDateBuilder::try_from_fields(fields, self, options)?;
```

Comment:
observation: this does not reject dates that aren't actually in the era. the constructors on `Date` still do though. is this intended?